src/Connectors/src/Connectors/RabbitMQ/RabbitMQHealthContributor.cs

```diff
@@ -4,110 +4,93 @@
+    private ConnectionInterfaceShim? _connectionInterfaceShim;
 
     public string Id { get; } = "RabbitMQ";
+    public string Host { get; }
+    public string? ServiceName { get; set; }
 
-    public RabbitMQHealthContributor(RabbitMQProviderConnectorFactory factory, ILogger<RabbitMQHealthContributor> logger = null)
+    public RabbitMQHealthContributor(object connectionFactory, string host, ILogger<RabbitMQHealthContributor> logger)
     {
-        _logger = logger;
-        _connFactory = factory.Create(null);
-    }
-
-    internal RabbitMQHealthContributor(object connection, string serviceName, string hostName, ILogger<RabbitMQHealthContributor> logger)
-    {
-        ArgumentGuard.NotNull(connection);
-        ArgumentGuard.NotNull(serviceName);
         ArgumentGuard.NotNull(logger);
+        ArgumentGuard.NotNullOrEmpty(host);
+        ArgumentGuard.NotNull(connectionFactory);
 
-        _connection = connection;
-        Id = serviceName;
-        _hostName = hostName;
+        _connectionFactoryInterfaceShim = new ConnectionFactoryInterfaceShim(RabbitMQPackageResolver.Default, connectionFactory);
+        Host = host;
         _logger = logger;
     }
 
-    public static IHealthContributor GetRabbitMQContributor(IConfiguration configuration, ILogger<RabbitMQHealthContributor> logger = null)
-    {
-        ArgumentGuard.NotNull(configuration);
-
-        Type rabbitMQImplementationType = RabbitMQTypeLocator.ConnectionFactory;
-
-        var info = configuration.GetSingletonServiceInfo<RabbitMQServiceInfo>();
-
-        var options = new RabbitMQProviderConnectorOptions(configuration);
-        var factory = new RabbitMQProviderConnectorFactory(info, options, rabbitMQImplementationType);
-        return new RabbitMQHealthContributor(factory, logger);
-    }
-
     public HealthCheckResult Health()
     {
-        _logger?.LogTrace("Checking RabbitMQ connection health");
-        var result = new HealthCheckResult();
+        _logger.LogTrace("Checking {DbConnection} health at {Host}", Id, Host);
 
-        if (!string.IsNullOrEmpty(_hostName))
+        var result = new HealthCheckResult
         {
-            result.Details.Add("host", _hostName);
+            Details =
+            {
+                ["host"] = Host
+            }
+        };
+
+        if (!string.IsNullOrEmpty(ServiceName))
+        {
+            result.Details["service"] = ServiceName;
         }
 
         try
         {
-            _connection ??= ReflectionHelpers.Invoke(RabbitMQTypeLocator.CreateConnectionMethod, _connFactory, null);
+            _connectionInterfaceShim ??= _connectionFactoryInterfaceShim.CreateConnection();
 
-            if (_connection == null)
+            if (!_connectionInterfaceShim.IsOpen)
             {
-                throw new ConnectorException("Failed to open RabbitMQ connection!");
+                throw new IOException("RabbitMQ connection is closed!");
```

Comment:
Doesn't `IOException` usually imply there's a file or disk-related issue?

Reply:
I searched for a while, but found no better candidate in the BCL exception types. I chose this one because Sonar doesn't like me throwing `Exception`, `ApplicationException`, or `SystemException` (too generic). However, I can suppress that rule in the code.

On a side note, I stepped away from throwing `ConnectorException` because it implies users can catch that and be fine. Which is definitely not the case in so many scenarios. We'd need to wrap all reflection calls in a try/catch, which takes away the possibility for clients to catch `TimeoutException`, but not `SqlServerException`.

Do you have a better exception type in mind?

Reply:
I think this was the original primary use case for `ConnectorException` -- something went wrong in Connectors that isn't covered by another common exception type. If you don't want to use `ConnectorException` in any of the new code, what do you think about an `InvalidOperationException` here?

Reply:
It doesn't literally say so, but based on the examples at https://learn.microsoft.com/en-us/dotnet/api/system.invalidoperationexception?view=net-7.0, `InvalidOperationException` usually indicates that _you_ (the developer) are doing something wrong. Instead of an intermittent, retriable failure, which is what we need here. While `IOException` isn't perfect, I still think it's the next best thing. It was proposed to add `ConnectionException` to the BCL, which [derives](https://github.com/dotnet/runtime/issues/39941) from `IOException`. And it got merged, but somehow it didn't survive.